Adding docs for integrating Istio with ZTWIM (SPIFFE/SPIRE) (#1319)
istio-ecosystem/sail-operator: docs/integrations/spire/spire.adoc

          image: ghcr.io/spiffe/spire-agent:1.5.1
          command: ["/opt/spire/bin/spire-agent"]
          args: [ "api", "watch",  "-socketPath", "/run/spire/sockets/socket" ]
          volumeMounts:
            - mountPath: /run/spire/sockets
              name: spiffe-workload-api
              readOnly: true
      volumes:
      - name: spiffe-workload-api
        csi:
          driver: csi.spiffe.io
          readOnly: true
EOF
----
Once the client pod is running try to fetch the SVID
[source,bash]
----
kubectl exec -it \
$(kubectl get pods -o=jsonpath='{.items[0].metadata.name}' -l app=client) \
 -- /opt/spire/bin/spire-agent api fetch -socketPath /run/spire/sockets/socket
----
If Spire was deployed and configured correctly
you should get something like this
[source,text]
----
Received 1 svid after 29.636075ms

SPIFFE ID:		spiffe://ocp.one/ns/default/sa/default
SVID Valid After:	 2025-10-21 14:04:03 +0000 UTC
SVID Valid Until:	 2025-10-21 15:04:13 +0000 UTC
CA #1 Valid After:	2025-10-21 07:38:03 +0000 UTC
CA #1 Valid Until:	2025-10-22 07:38:13 +0000 UTC
----
[[sail-operator]]
=== Sail Operator

[[instal-sail-operator]]
==== Install Sail Operator
[source,bash]
----

> helm repo add sail-operator https://istio-ecosystem.github.io/sail-operator

Reviewer: To avoid duplicate information of installation in different docs, you can point the user to the specific sail operator installation documentation: https://github.com/istio-ecosystem/sail-operator/blob/main/chart/README.md

Author: The installation doc does not instruct how to install Istio CNI. 
It's only cover the Helm Instal command. 
So, basically instead of specifying this 
```
helm repo add sail-operator https://istio-ecosystem.github.io/sail-operator
helm repo update
helm install sail-operator \
  sail-operator/sail-operator \
  -n istio-system \
  --create-namespace
  ``` 
I can point to https://github.com/istio-ecosystem/sail-operator/blob/main/chart/README.md, but it won't be good enough. 
Thus I'd suggest to duplicate the installation doc, so we'll make sure it is correct for the Spiffe integration.

helm repo update
helm install sail-operator \
  sail-operator/sail-operator \
  -nistio-system \
  --create-namespace
----
[[create-istio-cni]]
==== Create Istio CNI CR
[source,bash]
----
kubectl create namespace istio-cni
----
[source,bash]
----
cat <<EOF | kubectl apply -f -
apiVersion: sailoperator.io/v1
kind: IstioCNI
metadata:
  name: default
spec:
  namespace: istio-cni
EOF
----
[[create-istio-cr]]
==== Create Istio CR
Create `Istio` CR.

_If you are using Kubernetes 1.33 and have not disabled
support for native sidecars in the Istio control plane,
you must use initContainers in the injection template for sidecars.
This is required because native sidecar support changes
how sidecars are injected.
NOTE: The SPIRE injection template for gateways should
continue to use regular containers as before._
[source,bash]
----

> cat <<EOF | kubectl apply -f -

Reviewer: I'm running kind version 1.29 with default configuration, and I need to create the pod by running:
```
cat <<EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
  namespace: verify-istio-installation
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin
      version: v1
  template:
    metadata:
      annotations:
        inject.istio.io/templates: "sidecar,spireWithNativeSidecar"
      labels:
        app: httpbin
        version: v1
    spec:
      containers:
      - image: docker.io/mccutchen/go-httpbin:v2.15.0
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 8080
EOF
```

Maybe we should show the user how it is going to look in both cases: with and without native sidecars

    sidecarInjectorWebhook:
      templates:
        spireWithNativeSidecar: |
          spec:
            initContainers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
        spireWithoutNativeSidecar: |
          spec:
            containers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
            volumes:
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
EOF
----
_A note about `sidecarInjectorWebhook`.
Spiffe Workload API exposed over unix socket.
To avoid any host mounts we are using Spire CSI driver
which is securely injecting the workload api socket.
Thus, we must create sidecar injector template,
which will be responsible for mounting the Spire Agent

> socket as a volume to the envoy sidecar container._

Reviewer: ```suggestion
----

**About `sidecarInjectorWebhook`**:
_Spiffe Workload API exposed over unix socket.
To avoid any host mounts we are using Spire CSI driver
which is securely injecting the workload api socket.
Thus, we must create sidecar injector template,
which will be responsible for mounting the Spire Agent
socket as a volume to the envoy sidecar container._
```

Make sure the istiod up and running
[source,bash]
----
kubectl get deploy istiod -nistio-system
----
[[verify-istio-installation]]

> ==== Verify Istio Installation

Reviewer: ```suggestion
[[verify-istio-installation]]
==== Verify Istio Installation with Spire configuration
```

Note, in the inject template we are specifying `spire` template.
The spire injection template is responsible for mounting the Spiffe Workload API
socket into the sidecar container
[source,bash]
----
cat <<EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
  namespace: verify-istio-installation
spec:
  replicas: 1
  selector:
    matchLabels:
      app: httpbin
      version: v1
  template:
    metadata:
      annotations:
        inject.istio.io/templates: "sidecar,spireWithNativeSidecarWithNativeSidecar"
      labels:
        app: httpbin
        version: v1
    spec:
      containers:
      - image: docker.io/mccutchen/go-httpbin:v2.15.0
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 8080
EOF
----
Check that the workload identity was issued by SPIRE
[source,bash]
----
HTTPBIN_POD=$(kubectl get pod -l app=httpbin -nverify-istio-installation -o jsonpath="{.items[0].metadata.name}")
istioctl proxy-config secret "$HTTPBIN_POD" -nverify-istio-installation -o json | jq -r \
'.dynamicActiveSecrets[0].secret.tlsCertificate.certificateChain.inlineBytes' | base64 --decode > chain.pem

> openssl x509 -in chain.pem -text | grep SPIRE

Reviewer: ```suggestion
HTTPBIN_POD=$(kubectl get pod -l app=httpbin -n test -o jsonpath="{.items[0].metadata.name}")
istioctl proxy-config secret "$HTTPBIN_POD" -n test -o json | jq -r \
'.dynamicActiveSecrets[0].secret.tlsCertificate.certificateChain.inlineBytes' | base64 --decode > chain.pem
openssl x509 -in chain.pem -text | grep SPIRE
```

        imagePullPolicy: IfNotPresent
EOF
----
Currently, Istio configured with default PERMISSIVE mode.
Try to make http call without mTLS first
[source,bash]
----
CURL_POD=$(kubectl get pod -l app=curl -ntest-1 -o jsonpath="{.items[0].metadata.name}")
kubectl exec $CURL_POD -n test-1 -it -- curl -s -o /dev/null -w "%{http_code}" http://httpbin
----
You should get HTTP 200 status code. Now, lets enabled mTLS between two services.
We'll set `PeerAuthentication` to `STRICT` and will define two appropriate
`DestinationRules`
[source,bash]
----
cat <<EOF | kubectl apply -f -
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: default
  namespace: test-1
spec:
  mtls:
    mode: STRICT

---
apiVersion: networking.istio.io/v1
kind: DestinationRule
metadata:
  name: curl
  namespace: test-1
spec:
  host: curl
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
---
apiVersion: networking.istio.io/v1
kind: DestinationRule
metadata:
  name: httpbin
  namespace: test-1
spec:
  host: httpbin
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
---
EOF
----
Make the curl request again, you should get 200 response code.
[source,bash]
----

> CURL_POD=$(kubectl get pod -l app=curl -ntest-1 -o jsonpath="{.items[0].metadata.name}")

Reviewer: ```suggestion
CURL_POD=$(kubectl get pod -l app=curl -n test-1 -o jsonpath="{.items[0].metadata.name}")
```

apiVersion: v1
kind: Service
metadata:
  name: curl-service
  namespace: test-2
spec:
  selector:
    app: curl
  ports:
  - port: 80
    targetPort: 8080
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: curl
  namespace: test-2
spec:
  selector:
    matchLabels:
      app: curl
  template:
    metadata:
      labels:
        app: curl
    spec:
      containers:
      - name: curl
        image: curlimages/curl:8.16.0
        command:
        - /bin/sh
        - -c
        - |
            wget -O /tmp/spire.zip https://github.com/spiffe/spire/releases/download/v1.12.5/spire-1.12.5-linux-amd64-musl.tar.gz \
            && cd /tmp \
            && tar zxvf spire.zip \
            && mv /tmp/spire-1.12.5/bin/spire-agent /tmp/spire-agent \
            && sleep inf
        imagePullPolicy: IfNotPresent
        volumeMounts:
        - name: workload-socket
          mountPath: /tmp/spiffe-socket
          readOnly: true
      volumes:
      - name: workload-socket
        csi:
          driver: "csi.spiffe.io"
          readOnly: true
EOF
----
Make a direct HTTP request
from the external client to a service running inside the mesh.
[source,bash]
----
# get curl pod name

> CURL_POD=$(kubectl get pod -l app=curl -ntest-2 -o jsonpath="{.items[0].metadata.name}")

Reviewer: ```suggestion
CURL_POD=$(kubectl get pod -l app=curl -n test-2 -o jsonpath="{.items[0].metadata.name}")
```

The spire template is required to ensure the Spire Agent
socket is automatically mounted to
the Istio Ingress Gateway pod._

[source,bash]
----
# add istio helm repo
helm repo add istio https://istio-release.storage.googleapis.com/charts

# update the repo
helm repo update

# install the istio gateway helm chart
helm install istio-gateway -nistio-system \
  istio/gateway --set-json \
  'podAnnotations={"inject.istio.io/templates":"gateway,spireWithoutNativeSidecar"}'
----
Make sure the istio gateway is up and running
[source,bash]
----
kubectl get deploy istio-gateway -nistio-system
----
Create Istio Gateway CR for `httpbin` service in `test-1` namespace.

A note about the istio-gateway service:

This tutorial uses example.com as the placeholder domain. You should replace this with the correct domain for your setup.
You must configure DNS to resolve your domain to the gateway:

Cloud (e.g., AWS): If your cluster is in a cloud environment that provides a hostname (like an ELB), create a CNAME record mapping your domain to that hostname.

On-Premises/Bare-Metal: If your istio-gateway service has a LoadBalancerIP, create an A record mapping your domain to that external IP address.

> Alternative (nip.io): For quick testing, you can use nip.io. This method only works if your gateway service has an external IP address, not a CNAME.

Reviewer: ```suggestion
This tutorial uses example.com as the placeholder domain. You should replace this with the correct domain for your setup.
You must configure DNS to resolve your domain to the gateway:

* Cloud (e.g., AWS): If your cluster is in a cloud environment that provides a hostname (like an ELB), create a CNAME record mapping your domain to that hostname.
* On-Premises/Bare-Metal: If your istio-gateway service has a LoadBalancerIP, create an A record mapping your domain to that external IP address.
* Alternative (nip.io): For quick testing, you can use nip.io. This method only works if your gateway service has an external IP address, not a CNAME.
```

metadata:
  name: httpbin
  namespace: test-1
spec:
  hosts:
    - "httpbin.$BASE_DOMAIN"
  gateways:
    - httpbin-gateway
  http:
    - route:
      - destination:
          host: httpbin.test-1.svc.cluster.local
          port:
            number: 80
EOF
----

Make an http call to the httpbin service
[source,bash]
----
curl httpbin.$BASE_DOMAIN \
 -s -o /dev/null -w "%{http_code}"
----
If you receive an HTTP 200 code, it means your traffic
is being securely routed from the Istio Gateway pod
to the httpbin service using an mTLS connection.

[[bookinfo-app-x509-svid]]
=== Bookinfo app with SPIFFE
For the Bookinfo application, we will use the existing default namespace,
so there is no need to create a new one.
However, we must label the default namespace to enable
automatic Istio sidecar injection.
[source,bash]
----
kubectl label namespace default istio-injection=enabled
----

Deploy Bookinfo App from this xref:resources/bookinfo.yaml[manifest]
[source,bash]
----
kubectl create -f resources/bookinfo.yaml
----

Verify all Bookinfo workloads are up and running
[source,bash]
----
kubectl get deployment
----

Deploy Istio `VirtualService` and `Gateway` CRs for the Bookinfo App.

> Do not forget to export the `LB_IP` as mentioned previously

Reviewer: I think this is too vague. We should add a quick way for the user to get this, if not BASE_DOMAIN will be empty

Author: > I think this is too vague. We should add a quick way for the user to get this, if not BASE_DOMAIN will be empty

How you suggest to address this? 
Currently the `BASE_DOMAIN` defined  as `export BASE_DOMAIN=example.com`. 
This is pretty similar to how Istio docs asking to define variables. 

The bookinfo resources are here: docs/integrations/spire/resources/bookinfo.yaml

Reviewer: ```suggestion
Do not forget to export the `BASE_DOMAIN` as mentioned previously when we define base domain for the Gateway
```
